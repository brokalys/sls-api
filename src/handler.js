import { ApolloServer, ApolloError } from 'apollo-server-lambda';
import {
  ApolloServerPluginInlineTrace,
  ApolloServerPluginInlineTraceDisabled,
  ApolloServerPluginUsageReporting,
  ApolloServerPluginUsageReportingDisabled,
} from 'apollo-server-core';
import Buildings from './data-sources/buildings';
import Properties from './data-sources/properties';
import loadUser from './lib/auth';
import AuhtDirective from './lib/auth-directive';
import Bugsnag from './lib/bugsnag';
import mysql from './lib/db';
import schema from './schema/schema.graphql';
import dbConfig from './db-config';
import resolvers from './resolvers';
import './knex-extensions';

const isDevMode = process.env.STAGE === 'dev';
const isTestMode = process.env.NODE_ENV === 'test';

export const server = new ApolloServer({
  typeDefs: schema,
  schemaDirectives: {
    auth: AuhtDirective,
  },
  resolvers,
  dataSources: () => ({
    buildings: new Buildings(dbConfig),
    properties: new Properties(dbConfig),
  }),
  tracing: isDevMode,
  playground: isDevMode,
  context: async ({ event, req, request, context }) => {
    const { requestContext } = event ||
      req || { requestContext: { identity: {} } };

    return {
      user: await loadUser(requestContext.identity.apiKeyId),
      invokedFunctionArn: context ? context.invokedFunctionArn : '',
    };
  },
  formatError: (error) => {
    if (
      (isDevMode && !isTestMode) ||
      error instanceof ApolloError ||
      error.originalError instanceof ApolloError ||
      error.originalError === undefined
    ) {
      return error;
    }

    if (!isTestMode) {
      console.log(error);
      Bugsnag.notify(error);
    }

    return new Error('An unexpected error occurred. Please try again later.');
  },
  plugins: [
    isDevMode
      ? ApolloServerPluginInlineTrace()
      : ApolloServerPluginInlineTraceDisabled(),
    !isDevMode && !isTestMode
      ? ApolloServerPluginUsageReporting({
          sendVariableValues: {
            exceptNames: ['email', 'unsubscribe_key'],
          },
          logger: {
            debug: () => {},
            info: () => {},
            warn: console.warn,
            error: console.error,
          },
          generateClientInfo: ({ context, request }) => {
            const headers = request.http && request.http.headers;
            const clientVersion =
              headers['apollographql-client-version'] || '1.0.0';

            if (!context.user) {
              return {
                clientName: 'Unauthorized',
                clientVersion,
              };
            }

            return {
              clientName: context.user.apiKey.name,
              clientVersion,
            };
          },
        })
      : ApolloServerPluginUsageReportingDisabled(),
    {
      requestDidStart(requestContext) {
        return {
          willSendResponse() {
            return mysql.end();
          },
        };
      },
    },
  ],
});

exports.graphqlHandler = server.createHandler({
  cors: {
    origin: '*',
  },
});
