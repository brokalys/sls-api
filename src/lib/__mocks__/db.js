export default {
  connect: jest.fn(),
  getClient: jest.fn(() => ({
    escape: jest.fn(),
  })),
  query: jest.fn(() => []),
  end: jest.fn(),
};
