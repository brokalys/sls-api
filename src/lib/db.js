import serverlessMysql from 'serverless-mysql';

const mysql = serverlessMysql({
  config: {
    host: process.env.DB_HOST,
    database: process.env.DB_DATABASE,
    user: process.env.DB_USERNAME,
    password: process.env.DB_PASSWORD,
  },
  maxRetries: 5,
});

if (process.env.NODE_ENV !== 'test') {
  mysql.connect();
}

export default mysql;
