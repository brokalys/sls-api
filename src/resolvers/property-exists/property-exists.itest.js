import { createTestClient } from 'apollo-server-testing';

import { server } from 'handler';
import db from 'lib/db';

jest.mock('lib/db');

const mockInput = {
  foreign_id: 'id_123',
  source: 'ss.lv',
  type: 'sell',
  category: 'apartment',
  url: 'https://example.com',
  price: 10000,
  lat: 56.11,
  lng: 55.111112,
};

describe('Query: propertyExists', () => {
  let query;

  beforeEach(() => {
    const utils = createTestClient(server);
    query = utils.query;
  });

  test('checks if property exists with valid data and information in the DB', async () => {
    db.query.mockReturnValue([{ price: 1 }, { price: 2 }, { price: 3 }]);

    const response = await query({
      query: `
        query {
          propertyExists(
            source: "brokalys.com"
          )
        }
      `,
    });

    expect(response).toMatchSnapshot();
  });

  test('checks if property exists with all parameters and information in the DB', async () => {
    db.query.mockReturnValue([{ price: 1 }, { price: 2 }, { price: 3 }]);

    const response = await query({
      query: `
        query {
          propertyExists(
            source: "brokalys.com"
            foreign_id: "123"
            url: "https://brokalys.com"
            created_at: "2019-01-01T00:10:20"
          )
        }
      `,
    });

    expect(response).toMatchSnapshot();
  });

  test('checks if property exists with valid data and no information in the DB', async () => {
    db.query.mockReturnValue([]);

    const response = await query({
      query: `
        query {
          propertyExists(
            source: "brokalys.com"
          )
        }
      `,
    });

    expect(response).toMatchSnapshot();
  });

  test('fails if input source is invalid', async () => {
    const response = await query({
      query: `
        query {
          propertyExists(
            source: "malformed url"
            foreign_id: "id123"
          )
        }
      `,
    });

    expect(response).toMatchSnapshot();
  });

  test('fails if input foreign id is invalid', async () => {
    const response = await query({
      query: `
        query {
          propertyExists(
            source: "brokalys.com"
            foreign_id: "wrong ' 123"
          )
        }
      `,
    });

    expect(response).toMatchSnapshot();
  });

  test('fails if input contains unrecognized field', async () => {
    const response = await query({
      query: `
        query {
          propertyExists(
            source: "brokalys.com"
            foreign_id: "123"
            unkown: "field"
          )
        }
      `,
    });

    expect(response).toMatchSnapshot();
  });
});
