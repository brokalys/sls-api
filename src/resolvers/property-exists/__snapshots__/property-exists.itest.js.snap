// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query: propertyExists checks if property exists with valid data and information in the DB 1`] = `
Object {
  "data": Object {
    "propertyExists": true,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Query: propertyExists checks if property exists with valid data and no information in the DB 1`] = `
Object {
  "data": Object {
    "propertyExists": false,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Query: propertyExists fails if input contains unrecognized field 1`] = `
Object {
  "data": undefined,
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "GRAPHQL_VALIDATION_FAILED",
      },
      "locations": Array [
        Object {
          "column": 13,
          "line": 6,
        },
      ],
      "message": "Unknown argument \\"unkown\\" on field \\"Query.propertyExists\\".",
    },
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Query: propertyExists fails if input foreign id is invalid 1`] = `
Object {
  "data": null,
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "BAD_USER_INPUT",
        "details": Array [
          Object {
            "context": Object {
              "key": "foreign_id",
              "label": "foreign_id",
              "value": "wrong ' 123",
            },
            "message": "\\"foreign_id\\" must only contain alpha-numeric characters",
            "path": Array [
              "foreign_id",
            ],
            "type": "string.alphanum",
          },
        ],
        "exception": Object {
          "details": Array [
            Object {
              "context": Object {
                "key": "foreign_id",
                "label": "foreign_id",
                "value": "wrong ' 123",
              },
              "message": "\\"foreign_id\\" must only contain alpha-numeric characters",
              "path": Array [
                "foreign_id",
              ],
              "type": "string.alphanum",
            },
          ],
        },
      },
      "locations": Array [
        Object {
          "column": 11,
          "line": 3,
        },
      ],
      "message": "Input validation failed",
      "path": Array [
        "propertyExists",
      ],
    },
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Query: propertyExists fails if input source is invalid 1`] = `
Object {
  "data": null,
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "BAD_USER_INPUT",
        "details": Array [
          Object {
            "context": Object {
              "key": "source",
              "label": "source",
              "value": "malformed url",
            },
            "message": "\\"source\\" must contain a valid domain name",
            "path": Array [
              "source",
            ],
            "type": "string.domain",
          },
        ],
        "exception": Object {
          "details": Array [
            Object {
              "context": Object {
                "key": "source",
                "label": "source",
                "value": "malformed url",
              },
              "message": "\\"source\\" must contain a valid domain name",
              "path": Array [
                "source",
              ],
              "type": "string.domain",
            },
          ],
        },
      },
      "locations": Array [
        Object {
          "column": 11,
          "line": 3,
        },
      ],
      "message": "Input validation failed",
      "path": Array [
        "propertyExists",
      ],
    },
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
